File: data/src/app/transaction.kt
package data.app

import androidx.room.ColumnInfo
import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Entity
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.PrimaryKey
import androidx.room.Query
import androidx.room.Update
import kotlinx.coroutines.flow.Flow
import kotlinx.serialization.Serializable

@Entity(
    tableName = "transactions"
) @Serializable data class Transaction(
    @PrimaryKey val id: String,
    @ColumnInfo(name = "workspace_id") val workspaceId: String,
    @ColumnInfo(name = "user_id") val userId: String,
    val seq: Long,
    val operations: String,
    @ColumnInfo(name = "created_at", defaultValue = "CURRENT_TIMESTAMP") val createdAt: String = ""
)

@Dao interface TransactionDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE) suspend fun upsert(item: Transaction)
    @Update suspend fun update(item: Transaction)
    @Delete suspend fun delete(item: Transaction)

    @Query("SELECT * FROM transactions WHERE workspace_id = :workspaceId AND seq = 0 ORDER BY created_at ASC")
    suspend fun getPending(workspaceId: String): List<Transaction>

    @Query("SELECT * FROM transactions WHERE workspace_id = :workspaceId AND seq > :lastSeq ORDER BY seq ASC")
    suspend fun getSince(workspaceId: String, lastSeq: Long): List<Transaction>

    @Query("SELECT MAX(seq) FROM transactions WHERE workspace_id = :workspaceId")
    suspend fun getMaxSeq(workspaceId: String): Long?

    @Query("SELECT * FROM transactions WHERE id = :id")
    suspend fun getById(id: String): Transaction?

    @Query("SELECT * FROM transactions")
    fun getAll(): Flow<List<Transaction>>
}

File: server/data/app/transaction.ts
import { sqliteTable, text, integer, index } from "drizzle-orm/sqlite-core"
import { eq, sql } from "drizzle-orm"
import { DrizzleD1Database } from "drizzle-orm/d1"

export const Transaction = sqliteTable(
  "transactions",
  {
    created_at: text("created_at").default(sql`CURRENT_TIMESTAMP`),
    id: text("id").primaryKey(),
    operations: text("operations"),
    seq: integer("seq"),
    user_id: text("user_id"),
    workspace_id: text("workspace_id"),
  },
  (table) => ({
    idx0: index("transactions_workspace_id_seq_idx").on(table.workspace_id, table.seq),
  }),
)

export class TransactionDao {
  constructor(private db: DrizzleD1Database<any>) {}

  async upsert(item: typeof Transaction.$inferInsert) {
    return this.db
      .insert(Transaction)
      .values(item)
      .onConflictDoUpdate({ target: Transaction.id, set: item })
      .run()
  }

  async update(item: Partial<typeof Transaction.$inferInsert> & { id: any }) {
    return this.db.update(Transaction).set(item).where(eq(Transaction.id, item.id!)).run()
  }

  async delete(item: { id: any }) {
    return this.db.delete(Transaction).where(eq(Transaction.id, item.id!)).run()
  }

  async getById(id: string) {
    return this.db.select().from(Transaction).where(eq(Transaction.id, id)).get()
  }

  async getAll() {
    return this.db.select().from(Transaction).all()
  }

  async getPending(workspaceId: string) {
    return this.db
      .select()
      .from(Transaction)
      .where(sql`${Transaction.workspace_id} = ${workspaceId} AND ${Transaction.seq} = 0`)
      .all()
  }

  async getSince(workspaceId: string, lastSeq: number) {
    return this.db
      .select()
      .from(Transaction)
      .where(sql`${Transaction.workspace_id} = ${workspaceId} AND ${Transaction.seq} > ${lastSeq}`)
      .orderBy(Transaction.seq)
      .all()
  }

  async getMaxSeq(workspaceId: string) {
    const result = await this.db
      .select({ seq: Transaction.seq })
      .from(Transaction)
      .where(eq(Transaction.workspace_id, workspaceId))
      .orderBy(sql`${Transaction.seq} DESC`)
      .limit(1)
      .get()
    return result?.seq ?? null
  }
}
